monitor/statistics.cpp

```diff
@@ -190,27 +190,25 @@ statistics::statistics(monitor& mon, struct dnet_config *cfg) : m_monitor(mon)
 
 void statistics::add_provider(stat_provider *stat, const std::string &name)
 {
-	std::unique_lock<std::mutex> guard(m_provider_mutex);
-	m_stat_providers.emplace_back(std::unique_ptr<stat_provider>(stat), name);
+	boost::unique_lock<rw_lock> guard(m_provider_lock);
+	m_stat_providers.insert(make_pair(name, std::shared_ptr<stat_provider>(stat)));
 }
 
-struct provider_remover_condition
+void statistics::remove_provider(const std::string &name)
 {
-	std::string name;
+	boost::unique_lock<rw_lock> guard(m_provider_lock);
+	m_stat_providers.erase(name);
+}
 
-	bool operator() (const std::pair<std::unique_ptr<stat_provider>, std::string> &pair)
+std::shared_ptr<stat_provider> statistics::get_provider(const std::string &name)
```

Comment:
Do you really need this? On each event you will search provider by its name. It is extra costs which can be easy escaped.